### PR TITLE
lab1_glebosotov

### PR DESCRIPTION
**Pull request by glebosotov**
@azazkamaz @igooor-bb @Markovvn1 @yadariya @tasneem22

### CODE REVIEW

File: app_python/README.md [hunk 1]
```diff
@@ -0,0 +1,26 @@
+# GOVORIT MOSKVA
```
Comment (ghost) [inline app_python/README.md:1]: I like application name

File: app_python/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,6 @@
+FROM python:3.7
+COPY . /app
+WORKDIR /app
+RUN pip install -r requirements.txt
+ENTRYPOINT ["python3"]
+CMD ["app.py"]
```
Comment (markovvn1) [inline app_python/Dockerfile]: The assignment says to use a production ready framework to run the server (you are running application in a development server). You probably want to use something like [gunicorn](https://pypi.org/project/gunicorn/) or [waitress](https://flask.palletsprojects.com/en/2.2.x/tutorial/deploy/#run-with-a-production-server)

File: app_python/PYTHON.md [hunk 3]
```diff
@@ -0,0 +1,31 @@
+# Motivation and explanation of chosen stack
+
+## Best practices
+
+* add development files, that are nmot supposed to be pushed to .gitignore
+* use Docker for development and production
+* use virtualenv for development
```
Comment (markovvn1) [inline app_python/PYTHON.md:7]: You also use Jinja for templates, which is good practice

File: app_flutter/moscow_time/lib/main.dart [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+      body: StreamBuilder(
+        stream: Stream.periodic(const Duration(seconds: 1)),
+        builder: (context, snapshot) {
+          return Center(
+            child: Column(
+              mainAxisSize: MainAxisSize.min,
+              children: [
+                Text('Time in Moscow, Russia',
+                    style: Theme.of(context).textTheme.headline3),
+                Text(DateFormat('MM/dd/yyyy hh:mm:ss').format(DateTime.now()),
```
Comment (markovvn1) [inline app_flutter/moscow_time/lib/main.dart]: It seems like it has some problems with timezones. Right now is 09/04/2022 20:48:14 (not 08:48:14).
image

File: app_flutter/moscow_time/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,22 @@
+#Stage 1 - Install dependencies and build the app in a build environment
+FROM debian:latest AS build-env
```
Comment (AzazKamaz) [inline app_flutter/moscow_time/Dockerfile]: Better to use pre build flutter image. Building own each time takes too much time in production environment

File: app_flutter/docker-compose.yml [hunk 6]
```diff
@@ -0,0 +1,8 @@
+version: '2'
+services:
+  web:
+    build: moscow_time/
+    ports:
+      - "8080:80"
+    volumes:
+      - .moscow_time:/moscow_time
```
Comment (AzazKamaz) [inline app_flutter/docker-compose.yml]: Volume seems not to be necessary for production given there is just nginx which serves /usr/share/nginx/html

File: app_flutter/moscow_time/test/widget_test.dart [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+// gestures. You can also use WidgetTester to find child widgets in the widget
+// tree, read text, and verify that the values of widget properties are correct.
+
+import 'package:flutter/material.dart';
+import 'package:flutter_test/flutter_test.dart';
+
+import 'package:moscow_time/main.dart';
+
+void main() {
+  testWidgets('Counter increments smoke test', (WidgetTester tester) async {
```
Comment (AzazKamaz) [inline app_flutter/moscow_time/test/widget_test.dart]: Test not passing

File: app_python/app.py [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+@app.route('/')
+def display_time():
+    '''
+    Main function for default route.
+    Places time in the template file
+    '''
+    return render_template('template.html', datetime = str(datetime.now(tz=tz)))
+
+if __name__ == '__main__':
+    app.run(debug=True, host='0.0.0.0')
```
Comment (AzazKamaz) [inline app_python/app.py]: debug=True is not production ready probably

**Review by AzazKamaz — APPROVED** [review]

File: app_python/app.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+Simple Flask app for showing time in Moscow
+"""
+from datetime import datetime, timedelta, timezone
+from flask import Flask, render_template
+
+app = Flask(__name__)
+
+offset = timedelta(hours=3)
+tz = timezone(offset, name='MSK')
+
```
Comment (igooor-bb) [inline app_python/app.py:11]: I believe linter requires 2 empty lines according to pep8

File: app_python/app.py (hunk 8 above)
Comment (glebosotov) [reply app_python/app.py]: Resolved by using, `gunicorn` for running the server

File: app_flutter/moscow_time/Dockerfile (hunk 5 above)
Comment (glebosotov) [reply app_flutter/moscow_time/Dockerfile]: https://hub.docker.com/repository/docker/glebosotov/app_flutter_web

**Comment by glebosotov** [comment]
Updated with all mentioned problems fixed 🙌🏻

**Review by ghost — APPROVED** [review]

**Review by ghost — APPROVED** [review]

**Pull request closed without merging**